File: my-folder/0079-word-search/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        visit = set()
        def dfs(r, c, i):
            if i == len(word):
                return True
            
            if (r < 0 or c < 0 or r >= len(board) or c >= len(board[0]) or (r,c) in visit or word[i] != board[r][c]):
                return False

            visit.add((r,c))
            res = dfs(r + 1, c, i + 1) or dfs(r, c + 1, i + 1) or dfs(r - 1, c, i + 1) or dfs(r, c - 1, i + 1)
            visit.remove((r,c))
            return res
        
        for r in range(len(board)):
            for c in range(len(board[0])):
                if dfs(r, c, 0):
                    return True
        
        return False


# visit = (0,0), (0, 1) (0,2) (1, 2) 
# dfs(0, 0, 0)
# res = false or dfs(0, 1, 1) 
#                dfs(0, 2, 2) - dfs(1,2,3) - dfs()
                                        


                              
                                








    

        

        

File: my-folder/0079-word-search/solution.java
class Solution {
    public boolean exist(char[][] board, String word) {
        for(int m = 0; m < board.length; m++){
            for(int n = 0; n < board[0].length; n++){
                if(board[m][n] == word.charAt(0) && dfs(m, n, 0, word, board)){
                    return true;
                }
                }
            }

            return false;
        }
    

    private boolean dfs(int row, int col, int idx, String word, char[][] board){
        if(idx == word.length()){
            return true;
        }
        if(row < 0 || col < 0 || col >= board[0].length || row >= board.length){
            return false;
        }
        if(board[row][col] != word.charAt(idx)){
            return false;
        }
        
        char temp = board[row][col];
        board[row][col] = '#';

        boolean found = dfs(row + 1, col, idx + 1, word, board) || dfs(row - 1, col, idx + 1, word, board) || dfs(row, col + 1, idx + 1, word, board) || dfs(row, col - 1, idx + 1, word, board);

        board[row][col] = temp;

        return found;
    }
}

